[V3 Filter] Throw a filter Event on_filter_message_delete

entchen66 opened the pull request:
reference to https://github.com/Cog-Creators/Red-DiscordBot/pull/2443

### Type

- [ ] Bugfix
- [ ] Enhancement
- [x] New feature

### Description of the changes

Like Sinbad and Kowlin talked in PR 2443, I now add an event to the filter cog. It is thrown when a message got deleted and provide the message to custom cogs, just listen to the on_filter_message_delete event.

entchen66 committed:
Throw a filter Event on_filter_message_delete

reference to https://github.com/Cog-Creators/Red-DiscordBot/pull/2443

Tobotimus commented on redbot/cogs/filter/filter.py:
```diff
@@ -363,6 +363,7 @@ def __unload(self):
 
         if hits:
             try:
+                self.bot.dispatch("filter_message_delete", message)
```
The event should only be dispatched if the deletion succeeds. To do this, move this line down to 371, just after the `else:` line :smiley:

entchen66 replied:
oh yeah, sure^^

Tobotimus reviewed (changes requested):
Thanks for taking the time to open this PR. Just would like you to change it a bit :smile:

Tobotimus commented:
It might also be useful to provide `hits` as an argument to event listeners, so for example with your logging use case, the log message could display exactly which words/phrases were detected in the message.

entchen66 committed:
Update filter.py

moved line down, event will only fired, when deletion was succes

entchen66 committed:
Update filter.py

hits added to the event arguments. Useful for logging, so the evnt will provide the deleted message and the set of strings of wich where filtered.

Tobotimus commented on redbot/cogs/filter/filter.py:
```diff
@@ -364,6 +364,7 @@ def __unload(self):
         if hits:
             try:
                 await message.delete()
+                self.bot.dispatch("filter_message_delete", message, hits)
```
Please move it to line 371, after the `else:` line :D

entchen66 committed:
Update filter.py

moved out from the try statement

Tobotimus reviewed (approved):
Thanks for your contribution @entchen66 :smiley:
